src/net_processing.cpp

```diff
@@ -2710,8 +2709,15 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
         // set nodes not capable of serving the complete blockchain history as "limited nodes"
         pfrom.m_limited_node = (!(nServices & NODE_NETWORK) && (nServices & NODE_NETWORK_LIMITED));
 
-        {
+        // We only initialize the m_tx_relay data structure if:
+        // - this isn't an outbound block-relay-only connection; and
+        // - fRelay=true or we're offering NODE_BLOOM to this peer
+        //   (NODE_BLOOM means that the peer may turn on tx relay later)
+        const bool enable_tx_relay = !pfrom.IsBlockOnlyConn() &&
+                                      (fRelay || (pfrom.GetLocalServices() & NODE_BLOOM));
```

Comment:
I think that means peers to which we'd connect in `-blocksonly` mode, also wouldn't initialize `m_tx_relay` anymore. 
But `-blocksonly` mode until now allows to submit and relay transactions via RPC (`sendrawtransaction`), even if it that is obviously bad for privacy. 
With this change, attempting to send a tx via RPC in `-blocksonly` mode might lead to all of our peers disconnecting us because they wouldn't have `m_tx_relay` enabled anymore.

Reply:
Good catch. I've updated the `inv` and `tx` logic to explicitly test whether the peer is a block-relay-only connect instead of testing the existence of the Peer::TxRelay struct. That means that if we send a transaction to a peer while we're in blocksonly mode, then they will still process it.

Reply:
>  With this change, attempting to send a tx via RPC in -blocksonly mode might lead to all of our peers disconnecting us because they wouldn't have m_tx_relay enabled anymore.

Isn't there a test already that should check for this?

Reply:
>> With this change, attempting to send a tx via RPC in -blocksonly mode might lead to all of our peers disconnecting us because they wouldn't have m_tx_relay enabled anymore.

> Isn't there a test already that should check for this?

I would guess that:

https://github.com/bitcoin/bitcoin/blob/bdbabc50ba6c87ded97ea2bbacd3605c59cd12d0/test/functional/p2p_blocksonly.py#L43

would catch it. This branch will previously a draft since it was shared as a WIP, and I'm not sure if that test was failing.

Reply:
I did check that when I wrote my earlier comment: That line didn't fail because the peer we sent the tx to was just a python `P2PInterface()`, which wouldn't disconnect us for sending a tx like a full node would have.

Reply:
I guess the test could be changed to subclass `P2PInterface` and fail if the node under test sends a tx message.